Architecture Overview docs

ccfoley opened the pull request:
## Motivation
Document from an architectural perspective what Mobile Core does and what value it brings to the overall mobile solution.

## Description
The goal is to have a brief document on how the Mobile Core fits in with other key components of the overall architecture, i.e. Kubernetes, OpenShift, Service Catalog, Ansible Service Broker.
Another primary goal is to identify the role Mobile Core performs and the value add it brings. 

## Progress
- this is a first draft. Design and development is still ongoing with Mobile Core, therefore this doc will need to be revisited when design is near complete and stable.
- Reviewers; @maleck13 @david-martin 

## Additional Notes
- the source of the images should be added probably, unsure what file type is normally used for source diagrams?

wtrocki commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+=== Create a representation of a Mobile App
+The MC UI extension provides the user with the capabilitiy to create a representation of a Mobile App inside their Openshift project.
+
+MC allows the user to select the Mobile filter in the Service Catalog and select a representation of an App type to create. Note; an App representation is a set of basic properties for the App. It is not a running process. Currently its a visual represenation of an App.
```
We can also add info what mobile apps are for. For example:
`The mobile app is a resource that represents your mobile client app as part of the OpenShift UI. It gives us the context and information needed to show you relevant information around your particular mobile runtime as well as allowing us setup the different kind of mobile client builds required.`

From: https://github.com/aerogear/mobile-cli

david-martin replied:
category instead of filter

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+For mobile developers to be able to consume Mobile value add services, they must be available to them. The medium through which services are presented is the https://docs.openshift.com/container-platform/3.6/architecture/service_catalog/index.html[Openshift Service Catalog]. In order to achieve this, the MC must perform a number of tasks:
+
+** *Step 1.1*: Install / Configure an Ansible Service Broker
```
Since 3.7 the name is [OpenShift Ansible Broker](https://docs.openshift.com/container-platform/3.7/architecture/service_catalog/ansible_service_broker.html)

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
++
+image:images/5.x-Architecture-MC-ASB.png["My screenshot",width=1280, caption="Figure 1: Mobile Core Installation - Integrating Mobile Services into Service Catalog", title="Mobile Core Installation - Integrating Mobile Services into Service Catalog", link="images/5.x-Architecture-MC-ASB.png"]
++
+When the MC is installed, it will create a representation of a Broker inside the Service Catalog. In some cases a ASB may already exist. A service broker is a server that conforms to the https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md[OSB API] specification and manages a set of one or more services. The software could be hosted within your own OpenShift Container Platform ...
```
maybe "register with" instead of "create a representation of a Broker inside"

david-martin replied:
OAB instead of ASB

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+.Example of CRD for a Mobile App Representation
+====
+ MacBook-Pro:architecture user1$ kubectl describe crd mobileclients.mobile.k8s.io
```
minor thing. Maybe just `$` instead of `MacBook-Pro:architecture user1$`

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+The functionality provided through the OpenShift UI is also available on the command line. MC extends the Kubernetes and OpenShift CLI to provide mobile specific commands. More details in how to extend the base CLI is available https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/[here]
+
+All of the functionality available via the UI will be available from the CLI. The mobile CLI runs standalone aswell as through the _kubectl_ and _oc_ commands. E.g.
```
"as well"

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+image::images/5.x-ServiceCatalog-MobileApps.png[width=1280, title="Service Catalog - Mobile App Perspective", link="images/5.x-ServiceCatalog-MobileApps.png"]
+
+When this App representation is created, an instance of the associated CRD is created, see Example below.
```
Sounds like there's 2 things being created. "App representation" and "instance of the associated CRD".

Maybe just 'See below for an example of a Mobile App representation that uses the CRD'?

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+====
+Further details on creating a custom resource object (instantiation of the CRD) is detailed https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/[here].
+
+=== Instantiate a Mobile Service
```
Would 'Provision a Mobile Service' sound better?
I'm not sure what the correct terminology in OpenShift or the Catalog is.
I've heard 'Order' in the context of the Catalog. e.g. Order a Service.

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+Further details on creating a custom resource object (instantiation of the CRD) is detailed https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/[here].
+
+=== Instantiate a Mobile Service
+Another important stage in the Mobile App Dev process is to instantiate a Mobile value add Service to your project. This instantiation of a Mobile Service into you project is done via the Service Catalog, its not done through the MC, even though its the MCs UI extension is what allows you view a pick the service you want to instantiate.
```
The UI extension adds the Mobile category, but it's very much deferring to the Catalog UI in OpenShift for choosing the service and kicking off the provision.

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+image::images/5.x-ServiceCatalog-MobileServices.png[width=1280, title="Service Catalog - Mobile Service Perspective", link="images/5.x-ServiceCatalog-MobileServices.png"]
+
+What happens behind the scene is that the 'provision' ansible playbook of the service in question is invoked. This does the provisioning of the service into the project.
```
Is there a separate proposal to link too for this section? @maleck13 Did you write something about this approach before?

maleck13 replied:
https://github.com/aerogear/proposals/blob/master/apbs/create-secret-and-configmap-during-provision.md

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+image::images/5.x-Architecture-MC-ServiceInstantiation.png[width=1280, title="Instantiated/Provisioned Mobile Service", link="images/5.x-Architecture-MC-ServiceInstantiation.png"]
+
+The ConfigMap which is created contains public info about the mobile service. The info conatined in this ConfigMap is necessary for a remote/real Mobile Client/App to be able to interact with and consume the Mobile Service. It will be most likely be the Core Client SDK which will provide the Mobile Client with this configuration (TBC). The Secret that is created holds private info about the servi ...
```
Maybe a link to the SDK Proposal PR would be good here

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+** Example of https://github.com/aerogearcatalog/keycloak-apb/blob/master/roles/bind-keycloak-apb/tasks/main.yml[Keycloak Bind APB]
+* the MC would create a PodPreset for the consuming service.
+** A https://docs.openshift.com/container-platform/3.6/dev_guide/pod_preset.html[pod preset] is an object that injects user-specified information into pods as they are created.
+** The MC uses the pod preset concept to inject the secret created by the binding of the providing service into the consuming service (e.g. if you wanted to protect the Mobile Synch Service with authentication from Keycloak, the Keycloak secret would be injected into the Synch Service)
```
The MC doesn't do anything here.
Maybe something like this would read better?
"The Pod Preset defines how the Secret should be mounted into the consuming service"

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+* the bind APB is ran for providing service
+** Example of https://github.com/aerogearcatalog/keycloak-apb/blob/master/roles/bind-keycloak-apb/tasks/main.yml[Keycloak Bind APB]
+* the MC would create a PodPreset for the consuming service.
+** A https://docs.openshift.com/container-platform/3.6/dev_guide/pod_preset.html[pod preset] is an object that injects user-specified information into pods as they are created.
```
Might be good to link to kubernetes docs where possible, as the upstream may be more focused on Kubernetes.
https://kubernetes.io/docs/concepts/workloads/pods/podpreset/

david-martin commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+|Description
+
+|Keycloak
+|Feedhenry Synch Server
```
Sync tends to be used when shortening the name of the service

david-martin commented:
@ccfoley Might be good to add a terminology section and standardise on some of the terms we're using and what they mean (and if there's a difference). 
* Mobile Service, Mobile value add services, Mobile Clients, Mobile Apps, Mobile Synch Service, Synch Mobile Service, FeedHenry Sync(h) Server, Mobile Integrations, Mobile Service Integrations

Would be nice to use the same capitalisation throughout with each term then too e.g. Mobile Service vs. mobile Service.

wtrocki commented:
@ccfoley We can reuse extend existing documentation from here:
https://github.com/aerogear/mobile-cli

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+* *Step 1*: Ensure that Mobile Specific Services are made available through the Service Catalog
+
+For mobile developers to be able to consume Mobile value add services, they must be available to them. The medium through which services are presented is the https://docs.openshift.com/container-platform/3.6/architecture/service_catalog/index.html[Openshift Service Catalog]. In order to achieve this, the MC must perform a number of tasks:
```
> For mobile developers to be able to consume Mobile value add services
For mobile developers to be able to consume Mobile services

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,244 @@
+
+** *Step 1.1*: Install / Configure an Ansible Service Broker
++
+The Mobile value add services are represented as APBs (Ansible Playbooks). For these playbooks to be read in, and represented through the Service Catalog, an https://docs.openshift.com/container-platform/3.6/architecture/service_catalog/ansible_service_broker.html#service-catalog-spec-file[Ansible Service Broker] must be created.
```
Not a big fan of the wording: The Mobile value add services, I think maybe The Mobile focused services may come across better?

ccfoley commented:
With latest commit, I have tried to address all comments from @david-martin @wtrocki and @maleck13 . I have added a terminology section and just added definitions which are used on this page. We probably should create a single page, where we capture all terms across 5.x and make that the point of reference for terminology definitions.

finp commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+This is the Mobile Developers client application which will run on the Mobile device. It will interact with the Mobile Services which are provisioned via the SDKs.
+
+*Mobile App Representation*;
+A Mobile App Representation is a set of basic properties for the App. It is not a running process. Currently its a visual represenation of an App. It is a resource that represents your mobile client app as part of the OpenShift UI. It gives the context and information needed to show you relevant information around your particular mobile runtime as well as allowing the setup of different kind of m ...
```
I'm presuming this is the term that replaces 'mobile app client' ?

MAR defines the mobile application on OpenShift (eg MobileHR). You can access it using mobile cli to set and retrieve configuration properties (eg auth configuration). For example, to view all the configuration and context associated with your app:

mobile --namespace=MobileHR get clientconfig

ccfoley replied:
Am going to leave the definition as is Paul for now. I think it can be defined more clearly when the development has progressed further.

finp replied:
sure, just setting expectation for what we want to publish

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+****
+
+== Introduction
+The overall goal of a full Mobile solution is a group of services working together in order to solve a problem. Each of those component services offers a piece of the functionality necessary for the solution to work. But the overall solution needs all these component parts (i.e. services) to be integrated together. MC is the component which glues/integrates/orchestrates these component services.  ...
```
I think I would modify the wording here a little.

>The overall goal of a full Mobile solution is a group of services working together in order to solve a problem

```The overall goal of a full Mobile solution is to enable mobile developers to easily provision and consume a set of valuable mobile services on top of Kubernetes or OpenShift.  Each service solves an individual set of problems, but in many cases are designed to work together to provide a cohesive single solution```

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+
+** *Step 1.1*: Install / Configure an OpenShift Ansible Broker (OAB)
++
+The Mobile Services are represented as APBs (Ansible Playbooks). For these playbooks to be read in, and represented through the Service Catalog, an https://docs.openshift.com/container-platform/3.7/architecture/service_catalog/ansible_service_broker.html#service-catalog-spec-file[Ansible Service Broker] must be created.
```
APB link https://github.com/ansibleplaybookbundle/ansible-playbook-bundle

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
++
+image:images/5.x-Architecture-MC-ASB.png["My screenshot",width=1280, caption="Figure 1: Mobile Core Installation - Integrating Mobile Services into Service Catalog", title="Mobile Core Installation - Integrating Mobile Services into Service Catalog", link="images/5.x-Architecture-MC-ASB.png"]
++
+When the MC is installed, it will register a Broker with the Service Catalog. In some cases an OAB may already exist. A service broker is a server that conforms to the https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md[OSB API] specification and manages a set of one or more services. The software could be hosted within your own OpenShift Container Platform cluster or elsewhe ...
```
Service Class -> ClusterServiceClass https://github.com/kubernetes-incubator/service-catalog/blob/master/docs/resources.md

finp commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+The target audience for this document is developers who want a better understanding of what Mobile Core does, what value it brings and if they were to build a Mobile Service, how would they get MC to discover it and integrate with it.
+****
+
+== Terminology
```
what is 'mobile core'?

ccfoley replied:
Added a definition to the Terminology section

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+The MC as part of its startup procedure creates a CRD for representing Mobile Clients. At a later stage when a Mobile App representation is created, this CRD is instantiated.
+
+.Example of CRD for a Mobile App Representation
+====
```
I think it would be better to link to the actual CRD in the repo https://github.com/aerogear/mobile-cli/blob/master/artifacts/mobileclient_crd.yaml as it is likely to change over time

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+Further details on creating a custom resource object (instantiation of the CRD) is detailed https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/[here].
+
+=== Provision a Mobile Service
+Another important stage in the Mobile App Dev process is to provision a Mobile value add Service to your project. This provisioning of a Mobile Service into your project is done via the Service Catalog, its not done through the MC. The MC UI extension has provided the Mobile category for the user to find the services which can be used as part of a Mobile solution.
```
are we sticking with this terminology?  ```Mobile value add Service``` sounds strange and a bit contrived

ccfoley replied:
Thought that I had that changed back to 'Mobile Service'. We are sticking with the term 'Mobile Service'.

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,260 @@
+
+image::images/5.x-Architecture-MC-ServiceInstantiation.png[width=1280, title="Instantiated/Provisioned Mobile Service", link="images/5.x-Architecture-MC-ServiceInstantiation.png"]
+
+The ConfigMap which is created contains public info about the Mobile Service. The info conatined in this ConfigMap is necessary for a remote/real Mobile Client/App to be able to interact with and consume the Mobile Service. It will be most likely be the Core Client SDK which will provide the Mobile Client with this configuration (https://github.com/aerogear/proposals/pull/9/files[SDK Proposal]).  ...
```
The CLI will retrieve the configuration and the developer makes it available to the SDK via their project configuration

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,219 @@
++
+image:images/5.x-Architecture-MC-ASB.png["My screenshot",width=1280, caption="Figure 1: Mobile Core Installation - Integrating Mobile Services into Service Catalog", title="Mobile Core Installation - Integrating Mobile Services into Service Catalog", link="images/5.x-Architecture-MC-ASB.png"]
++
+When the MC is installed, it will register a Broker with the Service Catalog. In some cases an OAB may already exist. A service broker is a server that conforms to the https://github.com/openservicebrokerapi/servicebroker/blob/v2.13/spec.md[OSB API] specification and manages a set of one or more services. The software could be hosted within your own OpenShift Container Platform cluster or elsewhe ...
```
This paragraph is a little off. We switch between using the term OAB and ASB but these are actually the same things, To me using ASB throughout would make the most sense.
Also further down it says:
```
 Once the ASB is created, it will try and invoke a GET on the its configured endpoint, to retrieve the catalog of services it has access to.

```
Actually it is the service catalog that invokes the broker to return the services.


I think
```
At that point a user can see a representation of the services inside the Service Catalog.
```
should be ```At that point a user can see a representation of the services inside the Service Catalog UI.```

maleck13 commented on docs/architecture/architecture-overview.adoc:
```diff
@@ -0,0 +1,219 @@
+
+* *Step 4*: Install the Mobile CLI
+
+The functionality provided through the OpenShift UI is also available on the command line. MC extends the Kubernetes and OpenShift CLI to provide mobile specific commands. More details in how to extend the base CLI is available https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/[here]
```
``` More details in how to ``` -> ```More details on how to```

ccfoley commented:
@maleck13 Addressed your two comments.
